### PR TITLE
add `lexproduct`, lexicographic-order product iterator

### PR DESCRIPTION
**Pull request by JeffBezanson**

### CODE REVIEW

File: base/iterator.jl [hunk 1]
```diff
@@ -555,7 +562,7 @@ a tuple whose `i`th element comes from the `i`th argument iterator. The first it
 changes the fastest. Example:
 
 ```jldoctest
-julia> collect(product(1:2,3:5))
+julia> vec(collect(product(1:2,3:5)))
```
Comment (tkelman) [inline base/iterator.jl]: run `make docs` and commit the result

Comment (JeffBezanson) [inline base/iterator.jl]: Ok, I'll do that. I'll go with `IterTools` and see what people think.

**Comment by tkelman** [comment]
I assume this is something you're planning on using for some other feature, which is why it needs to be here rather than Iterators.jl?

**Comment by JeffBezanson** [comment]
Not necessarily, they just seem to go together. It also reuses some internal types and functions in Base defined for `product`, so it's much easier to implement here.

**Pull request closed without merging**